File: src/main/java/com/example/cqrs/command/entity/event/AccountCreatedEvent.java
package com.example.cqrs.command.entity.event;

import com.example.cqrs.command.entity.event.metadata.EventMetadata;
import jakarta.persistence.DiscriminatorValue;
import jakarta.persistence.Entity;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;

@NoArgsConstructor(access = AccessLevel.PROTECTED)
@DiscriminatorValue("AccountCreatedEvent")
@Entity
public class AccountCreatedEvent extends AbstractAccountEvent {

    public AccountCreatedEvent(String accountId, LocalDateTime eventDate,
                             double amount, EventMetadata metadata) {
        super(accountId, eventDate, amount, metadata);
    }

    // factory method
    public static AccountCreatedEvent of(String accountId, LocalDateTime eventDate, double amount, EventMetadata metadata) {
        return new AccountCreatedEvent(accountId, eventDate, amount, metadata);
    }

}

File: src/main/java/com/example/cqrs/command/entity/event/account/AccountCreatedEvent.kt
package com.example.cqrs.command.entity.event.account

import com.example.cqrs.command.entity.event.base.AccountEvent
import com.example.cqrs.command.entity.event.metadata.EventMetadata
import jakarta.persistence.DiscriminatorValue
import jakarta.persistence.Entity
import java.time.LocalDateTime

@DiscriminatorValue("AccountCreatedEvent")
@Entity
class AccountCreatedEvent(
    accountId: String,
    amount: Double? = 0.0,
    eventDate: LocalDateTime,
    metadata: EventMetadata
) : AccountEvent(
    id = null,
    accountId = accountId,
    amount = amount,
    eventDate = eventDate,
    metadata = metadata
) {
    companion object {
        fun of(
            accountId: String,
            eventDate: LocalDateTime,
            amount: Double = 0.0,
            metadata: EventMetadata
        ): AccountCreatedEvent {
            return AccountCreatedEvent(
                accountId = accountId,
                amount = amount,
                eventDate = eventDate,
                metadata = metadata
            )
        }
    }
}